Firestore/core/test/firebase/firestore/remote/grpc_stream_test.cc

```diff
@@ -0,0 +1,318 @@
+    observed_states.push_back("OnStreamWrite");
+  }
+  void OnStreamError(const util::Status& status) override {
+    observed_states.push_back("OnStreamError");
+  }
+
+  int generation() const override {
+    return gen;
+  }
+
+  std::vector<std::string> observed_states;
+  int gen = 0;
+};
+
+class GrpcStreamTest : public testing::Test {
+ public:
+  enum OperationResult { Ok, Error, Ignore };
+
+  GrpcStreamTest() {
+    grpc::GenericStub grpc_stub{grpc::CreateChannel(
+        "", grpc::SslCredentials(grpc::SslCredentialsOptions()))};
+
+    auto grpc_context_owning = absl::make_unique<grpc::ClientContext>();
+    grpc_context = grpc_context_owning.get();
+    auto grpc_call = grpc_stub.PrepareCall(grpc_context_owning.get(), "",
+                                           grpc_queue.queue());
+    observer = absl::make_unique<Observer>();
+    stream = GrpcStream::MakeStream(std::move(grpc_context_owning),
+                                    std::move(grpc_call), observer.get(),
+                                    &grpc_queue);
+  }
+
+  // This is a very hacky way to simulate GRPC finishing operations without
+  // actually connecting to the server: cancel the stream, which will make the
+  // operation fail fast and be returned from the completion queue, then
+  // complete the operation. It relies on `ClientContext::TryCancel` not
+  // complaining about being invoked more than once.
+  void ForceFinish(std::initializer_list<OperationResult> results) {
+    grpc_context->TryCancel();
+
+    for (OperationResult result : results) {
+      bool ignored_ok = false;
+      // TODO(varconst): use a timeout, otherwise this might block if there's
+      // a bug.)
+      GrpcOperation* operation = grpc_queue.Next(&ignored_ok);
+      ASSERT_NE(operation, nullptr);
+      if (result == OperationResult::Ok) {
+        operation->Complete(true);
+      } else if (result == OperationResult::Error) {
+        operation->Complete(false);
+      }
+      // Otherwise, the operation is ignored.
+    }
+  }
+  void ForceFinishAndShutdown(std::initializer_list<OperationResult> results) {
+    ForceFinish(results);
+
+    grpc_queue.Shutdown();
+    bool unused_ok = false;
+    GrpcOperation* should_be_null = grpc_queue.Next(&unused_ok);
+    EXPECT_EQ(should_be_null, nullptr);
+  }
+
+  // This is to make `EXPECT_EQ` a little shorter and work around macro
+  // limitations related to initializer lists.
+  std::vector<std::string> States(std::initializer_list<std::string> states) {
+    return {states};
+  }
+
+  grpc::ClientContext* grpc_context = nullptr;
+  GrpcCompletionQueue grpc_queue;
+  std::unique_ptr<Observer> observer;
+  std::shared_ptr<GrpcStream> stream;
+};
+
+TEST_F(GrpcStreamTest, CannotStartTwice) {
+  EXPECT_NO_THROW(stream->Start());
+  EXPECT_ANY_THROW(stream->Start());
+}
+
+TEST_F(GrpcStreamTest, CannotWriteBeforeStreamIsOpen) {
+  EXPECT_ANY_THROW(stream->Write({}));
+  stream->Start();
+  EXPECT_ANY_THROW(stream->Write({}));
```

Comment:
Is this relying on the state is not immediately open after call of stream->Start()? I am thinking of flakiness.

Reply:
Stream won't become open until `StreamStart` operation completes; in these tests, it won't complete without a call to `ForceFinish`. So the stream is reliably in `Starting` state at this point.